modules/ingester/instance.go

```diff
@@ -267,6 +269,15 @@ func (i *instance) CutBlockIfReady(maxBlockLifetime time.Duration, maxBlockBytes
 
 		completingBlock := i.headBlock
 
+		// Now that we are adding a new block take the blocks mutex.
+		// A warning about deadlocks!!  This area does a hard-acquire of both mutexes.
+		// To avoid deadlocks this function and all others must acquire them in
+		// the ** same_order ** or else. i.e. another function can't acquire blocksMtx
```

Comment:
not enough !!! after "or else".

how confident are we that "TestInstanceDoesNotRace" will catch deadlocks?

Reply:
Maybe you were joking but more !!! added.    Actually the deadlocks were being caught by `TestInstanceSearchDoesNotRace` because it exercises more read path (search, tags, tag values). I tested with count=10 and it virtually always triggered until I fixed them.  I think just one time it didn't.  So probably nearly 100% within 20 runs?   Our CI runs often enough it should be good, and the failure mode will be a timeout in one of these DoesNotRace tests.